gobblin-core/src/main/java/gobblin/writer/RetryWriter.java

```
@@ -0,0 +1,162 @@
+
+  private final DataWriter<D> writer;
+  private final Retryer<Void> retryer;
+  private boolean isFailure = false;
+
+  public RetryWriter(DataWriter<D> writer, State state) {
+    this.writer = writer;
+    this.retryer = buildRetryer(state);
+  }
+
+  /**
+   * Build Retryer.
+   * - If Writer implements Retriable, it will use the RetryerBuilder from the writer.
+   * - Otherwise, it will use DEFAULT writer builder.
+   *
+   * - If Gobblin metrics is enabled, it will emit all failure count in to metrics.
+   *
+   * @param state
+   * @return
+   */
+  private Retryer<Void> buildRetryer(State state) {
+    RetryerBuilder<Void> builder = null;
+    if (writer instanceof Retriable) {
+      builder = ((Retriable) writer).getRetryerBuilder();
+    } else {
+      builder = getDefaultRetryBuilder();
+    }
+
+    if (GobblinMetrics.isEnabled(state)) {
+      final Optional<Meter> retryMeter = Optional.of(Instrumented.getMetricContext(state, getClass()).meter(FAILED_RETRY_WRITES_METER));
+
+      builder.withRetryListener(new RetryListener() {
+        @Override
+        public <V> void onRetry(Attempt<V> attempt) {
+          if (attempt.hasException()) {
+            LOG.warn("Caught exception. This may be retried.", attempt.getExceptionCause());
+            Instrumented.markMeter(retryMeter);
+          }
+        }
+      });
+    }
+    return builder.build();
+  }
+
+  @Override
+  public void close() throws IOException {
+    writer.close();
+  }
+
+  @Override
+  public void write(final D record) throws IOException {
+    //Need a Callable interface to be wrapped by Retryer.
+    Callable<Void> writeCall = new Callable<Void>() {
+      @Override
+      public Void call() throws Exception {
+        writer.write(record);
+        return null;
+      }
+    };
+
+    try {
+      this.retryer.wrap(writeCall).call();
+    } catch (Exception e) {
+      isFailure = true;
+      throw new RuntimeException(e);
```

Synced offline.
Will remove "isFailure" logic as commit won't happen if write fails. (Write needs to throw exception to fail fast.)